integration-tests/src/tests/network/runner.rs

```diff
@@ -688,95 +596,70 @@ impl Runner {
         self.ports = Some(ports);
         self.validators = Some(validators);
 
-        let pm_addr: Vec<_> = self
-            .test_config
-            .iter()
-            .enumerate()
-            .map(|(node_id, _)| self.setup_node(node_id))
-            .collect();
-
-        RunningInfo { pm_addr, peers_info }
+        let mut nodes = vec![];
+        for (node_id, _) in self.test_config.iter().enumerate() {
+            nodes.push(Some(self.setup_node(node_id).await?));
+        }
+        Ok(RunningInfo { runner: self, nodes, peers_info })
     }
 }
 
-/// Use to start running the test.
+/// Executes the test.
 /// It will fail if it doesn't solve all actions.
-pub fn start_test(runner: Runner) {
-    run_actix(async {
-        runner.start();
-    })
-}
-
-impl Actor for Runner {
-    type Context = Context<Self>;
-
-    fn started(&mut self, ctx: &mut Self::Context) {
-        let addr = ctx.address();
-
-        init_test_logger();
-
-        let info = self.build();
-
-        let mut pointer = None;
-        let mut flag = Arc::new(AtomicBool::new(true));
-        let mut state_machine = self.state_machine.take().unwrap();
-        self.info = Some(Arc::new(RwLock::new(info)));
-
-        let info = self.info.as_ref().cloned().unwrap();
-
-        let can_write_log = Arc::new(AtomicBool::new(true));
-        WaitOrTimeoutActor::new(
-            Box::new(move |ctx| {
-                if can_write_log.swap(false, Ordering::Relaxed) == true {
-                    debug!(target: "network", "runner.rs: WaitOrTimeoutActor");
+/// start_test will block until test is complete.
+pub fn start_test(runner: Runner) -> anyhow::Result<()> {
+    init_test_logger();
+    let r = tokio::runtime::Runtime::new()?;
+    r.block_on(async {
+        let mut info = runner.build().await?;
+        let actions = std::mem::replace(&mut info.runner.state_machine.actions, vec![]);
+        let actions_count = actions.len();
+
+        let timeout = time::Duration::from_secs(15);
+        let step = time::Duration::from_millis(10);
+        let start = time::Instant::now();
+        for (i, a) in actions.into_iter().enumerate() {
+            loop {
+                if time::Instant::elapsed(&start) >= timeout {
+                    bail!("timeout while executing action {}/{}", i, actions_count)
                 }
-                if flag.load(Ordering::Relaxed) {
-                    pointer = Some(pointer.map_or(0, |x| x + 1));
-                    flag = Arc::new(AtomicBool::new(false));
+                if a(&mut info).await? {
```

Comment:
I wonder if we should add timeout here? Feels like, if we hit deadlock, we'd hit it *inside* action?

Reply:
done